notebooks/data_preparation_and_analysis.ipynb

```diff
@@ -1,226 +1,319 @@
+    "\n",
+    "# Balance recordings per device\n",
+    "balanced_device_files = balance_recordings_by_device(wav_files, max_files_per_device=2)\n",
+    "print(\"\\nFinal Balanced Device Files:\", balanced_device_files)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 2. Statistics about authorized/unauthorized speakers"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 21,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Authorized speakers recordings:\n",
+      "Statistics:\n",
+      "        Total files: 30,\n",
+      "        Total duration: 4978.07 sec,\n",
+      "        Average duration: 165.94 sec, \n",
+      "        Duration range: 132.03 - 198.58 sec\n",
+      "        \n",
+      "\n",
+      "Unauthorized speakers recordings:\n",
+      "Statistics:\n",
+      "        Total files: 70,\n",
+      "        Total duration: 11118.22 sec,\n",
+      "        Average duration: 158.83 sec, \n",
+      "        Duration range: 116.16 - 224.20 sec\n",
+      "        \n"
+     ]
+    }
+   ],
+   "source": [
+    "authorized_speakers_files = []\n",
+    "unauthorized_speakers_files = []\n",
+    "\n",
+    "for file in wav_files:\n",
+    "    speaker_id = os.path.split(file)[-1].split('_')[0]\n",
+    "    if speaker_id in VALID_ACCESS_LABELS:\n",
+    "        authorized_speakers_files.append(file)\n",
+    "    else:\n",
+    "        unauthorized_speakers_files.append(file)\n",
+    "\n",
+    "print(\"Authorized speakers recordings:\")\n",
+    "soa_authorized = SOAAudioClips(authorized_speakers_files)\n",
+    "print(duration_statistics(soa_authorized.clips))\n",
+    "\n",
+    "print(\"\\nUnauthorized speakers recordings:\")\n",
+    "soa_unauthorized = SOAAudioClips(unauthorized_speakers_files)\n",
+    "print(duration_statistics(soa_unauthorized.clips))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 3. Split files into train, validation, and test sets"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "train_files, val_files, test_files = prepare_datasets(DATA_DIR)\n",
+    "print(f\"Training files: {len(train_files)} | Validation files: {len(val_files)} | Test files: {len(test_files)}\")\n",
+    "\n",
+    "def save_spectrogram(spectrogram, output_path):\n",
+    "    plt.imsave(output_path, spectrogram, cmap='gray')\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 4. Calculate and display statistics about raw dataset"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "soa_train_full_clips = SOAAudioClips(train_files)\n",
+    "soa_test_full_clips = SOAAudioClips(test_files)\n",
+    "soa_val_full_clips = SOAAudioClips(val_files)\n",

```

Comment:
I want `train_files` to be balanced to check how it improves the training process. And I think that it should be balanced in both ways user + device, not only one of them at the time. Also, the authorized/unauthorized ratio is to be 1:1 in the training dataset. 
example:
Our dataset:
```
speaker: f1, recordings 1-8, devices {A, B}
speaker: m1, recordings 1-8, devices {A, B}
speaker: m2, recordings 1-8, devices {A, B}
```
Authorized: `{ f1 }`
Unauthorized: `{ m1, m2 }`
I think that a reasonable balance of every aspect of this dataset would be to:
1. m1, m2 should have 2 times fewer recordings than f1 because the unauthorized set size is 2 times greater.
2. When choosing a recording to be in train dataset, we should include this recording from every available device. We don't want common recordings in train/validation/test datasets to prevent testing on training data.

According to this I would split the example dataset into for example:
Train: `(f1, 1, A), (f1, 1, B),(f1, 2, A), (f1, 2, B),(f1, 3, A), (f1, 3, B),(f1, 4, A), (f1, 4, B), (m1, 1, A), (m1, 1, B), (m1, 2, A), (m1, 2, B), (m2, 1, A), (m2, 1, B), (m2, 2, A), (m2, 2, B)`
Validation and Test - I would try to make the same distribution of recordings as in the input dataset (we want to test a real-world scenario) but from the files not included in `train_files`. And remember, about point 2, we don't want to ex. (m1,7,A) in test and (m1,7,B) in validation because these are recordings of the same thing, just different noises.

What do you think? Does this make sense to you?

Reply:
I have managed to split the data (from all catalogs (except cleanraw, clean, production) in the following way:

`Train Dataset Statistics: {'Total Files': 378, 'Authorized': 194, 'Unauthorized': 184, 'Speakers': Counter({'m8': 35, 'f8': 34, 'm6': 32, 'm3': 31, 'f1': 31, 'f7': 31, 'f6': 19, 'm1': 17, 'f5': 17, 'm10': 14, 'f9': 14, 'm7': 14, 'm9': 14, 'f4': 13, 'f3': 13, 'f10': 13, 'm2': 11, 'm5': 11, 'm4': 10, 'f2': 4}), 'Devices': Counter({'ipad': 243, 'iphone': 135}), 'Rooms': Counter({'balcony1': 88, 'livingroom1': 81, 'bedroom1': 80, 'confroom2': 51, 'office1': 43, 'confroom1': 35})}

Validation Dataset Statistics: {'Total Files': 10, 'Authorized': 5, 'Unauthorized': 5, 'Speakers': Counter({'f1': 2, 'f2': 1, 'm6': 1, 'm2': 1, 'f10': 1, 'f7': 1, 'm10': 1, 'm3': 1, 'f3': 1}), 'Devices': Counter({'ipad': 6, 'iphone': 4}), 'Rooms': Counter({'bedroom1': 4, 'balcony1': 2, 'livingroom1': 2, 'confroom2': 1, 'office1': 1})}

Test Dataset Statistics: {'Total Files': 10, 'Authorized': 6, 'Unauthorized': 4, 'Speakers': Counter({'f8': 3, 'f1': 2, 'f3': 2, 'f2': 1, 'm8': 1, 'f5': 1}), 'Devices': Counter({'ipad': 8, 'iphone': 2}), 'Rooms': Counter({'confroom2': 3, 'balcony1': 2, 'bedroom1': 2, 'livingroom1': 1, 'office1': 1, 'confroom1': 1})}
`

Compared to our regular split from example folder like ipad_balcony1

`
Train Dataset Statistics: {'Total Files': 64, 'Authorized': 19, 'Unauthorized': 45, 'Speakers': Counter({'f5': 5, 'm1': 4, 'f3': 4, 'f1': 4, 'm10': 4, 'm9': 4, 'm7': 4, 'f9': 4, 'f8': 4, 'm2': 4, 'm6': 3, 'm4': 3, 'm3': 3, 'm8': 3, 'f7': 2, 'm5': 2, 'f4': 2, 'f2': 2, 'f6': 2, 'f10': 1}), 'Devices': Counter({'ipad': 64}), 'Rooms': Counter({'balcony1': 64})}

Validation Dataset Statistics: {'Total Files': 16, 'Authorized': 6, 'Unauthorized': 10, 'Speakers': Counter({'f4': 2, 'f10': 2, 'f7': 2, 'm3': 2, 'm1': 1, 'f1': 1, 'm2': 1, 'f2': 1, 'm8': 1, 'm7': 1, 'm9': 1, 'm5': 1}), 'Devices': Counter({'ipad': 16}), 'Rooms': Counter({'balcony1': 16})}

Test Dataset Statistics: {'Total Files': 20, 'Authorized': 5, 'Unauthorized': 15, 'Speakers': Counter({'f6': 3, 'm6': 2, 'm4': 2, 'f2': 2, 'f10': 2, 'm5': 2, 'm10': 1, 'f9': 1, 'f8': 1, 'f7': 1, 'f4': 1, 'f3': 1, 'm8': 1}), 'Devices': Counter({'ipad': 20}), 'Rooms': Counter({'balcony1': 20})}`

gives a decent difference :)